gdal/frmts/sigdem/sigdemdataset.cpp

```diff
@@ -0,0 +1,626 @@
+                "frmt_various.html#SIGDEM");
+        poDriver->SetMetadataItem(GDAL_DMD_EXTENSION, "sigdem");
+
+        poDriver->SetMetadataItem(GDAL_DCAP_VIRTUALIO, "YES");
+        poDriver->pfnOpen = SIGDEMDataset::Open;
+        poDriver->pfnCreateCopy = SIGDEMDataset::CreateCopy;
+
+        GetGDALDriverManager()->RegisterDriver(poDriver);
+    }
+}
+
+static int32_t GetCoordinateSystemId(const char* pszProjection) {
+    int32_t coordinateSystemId = 0;
+    OGRSpatialReference* poSRS = BuildSRS(pszProjection);
+    if (poSRS != nullptr) {
+        std::string pszRoot;
+        if (poSRS->IsProjected()) {
+            pszRoot = "PROJCS";
+        } else {
+            pszRoot = "GEOCS";
+        }
+        const char *pszAuthName = poSRS->GetAuthorityName(pszRoot.c_str());
+        const char *pszAuthCode = poSRS->GetAuthorityCode(pszRoot.c_str());
+        if (pszAuthName != nullptr && EQUAL(pszAuthName, "EPSG")
+                && pszAuthCode != nullptr) {
+            coordinateSystemId = atoi(pszAuthCode);
+        }
+    }
+    delete poSRS;
+    return coordinateSystemId;
+}
+
+SIGDEMDataset::SIGDEMDataset(SIGDEMHeader& sHeaderIn) :
+        fpImage(nullptr),
+        bGotTransform(false),
+        pszProjection(CPLStrdup("")),
+        sHeader(sHeaderIn) {
+    this->nRasterXSize = sHeader.nCols;
+    this->nRasterYSize = sHeader.nRows;
+
+    this->bGotTransform = true;
+    this->adfGeoTransform[0] = sHeader.dfMinX;
+    this->adfGeoTransform[1] = sHeader.dfXDim;
+    this->adfGeoTransform[2] = 0.0;
+    this->adfGeoTransform[3] = sHeader.dfMaxY;
+    this->adfGeoTransform[4] = 0.0;
+    this->adfGeoTransform[5] = -sHeader.dfYDim;
+}
+
+SIGDEMDataset::~SIGDEMDataset() {
+    FlushCache();
+
+    if (fpImage != nullptr) {
+        if (VSIFCloseL(fpImage) != 0) {
+            CPLError(CE_Failure, CPLE_FileIO, "I/O error");
+        }
+    }
+
+    CPLFree(pszProjection);
+}
+
+GDALDataset* SIGDEMDataset::CreateCopy(
+    const char * pszFilename,
+    GDALDataset * poSrcDS,
+    CPL_UNUSED int bStrict,
+    CPL_UNUSED char ** papszOptions,
+    GDALProgressFunc pfnProgress,
+    void * pProgressData) {
+    const int nBands = poSrcDS->GetRasterCount();
+    double adfGeoTransform[6] = { };
+    if (poSrcDS->GetGeoTransform(adfGeoTransform) != CE_None) {
+        CPLError(CE_Failure, CPLE_NotSupported,
+                "SIGDEM driver requires a valid GeoTransform.");
+        return nullptr;
+    }
+
+    if (nBands != 1) {
+        CPLError(CE_Failure, CPLE_NotSupported,
+                "SIGDEM driver doesn't support %d bands.  Must be 1 band.",
+                nBands);
+        return nullptr;
+    }
+
+    VSILFILE *fp = VSIFOpenL(pszFilename, "wb");
+    if (fp == nullptr) {
+        CPLError(CE_Failure, CPLE_OpenFailed,
+                "Attempt to create file `%s' failed.", pszFilename);
+        return nullptr;
+    }
+
+    GDALRasterBand* band = poSrcDS->GetBands()[0];
```

Comment:
you should check poSrcDS->GetRasterCount() == 0, since that can occur, and is the likely cause for the crash in misc_6 (since it tests that case)

Reply:
See line 169 & 177

```c++
const int nBands = poSrcDS->GetRasterCount();
  :
    if (nBands != 1) {
        CPLError(CE_Failure, CPLE_NotSupported,
                "SIGDEM driver doesn't support %d bands.  Must be 1 band.",
                nBands);
        return nullptr;
    }
```

Reply:
Ah indeed. I overlooked this